### ISSUE TITLE
Partial derivates are returned in the wrong order with Adjoint+Catalyst

### ISSUE THREAD

**Issue by dime10** [PennyLaneAI/pennylane-lightning#1335]
I'm tracking down a bug which gives incorrect results when differentiating multiple expectation values with the adjoint method in Catalyst. From my investigation the right numerical values are all there but they are not returned in the correct grouping and order.

Minimal example (requires [this branch](https://github.com/PennyLaneAI/catalyst/pull/2428) since we had unintentionally placed an assertion preventing multiple qnode results):
```py
import jax
import jax.numpy as jnp
import pennylane as qml
from pennylane import numpy as np

from catalyst import *

n_wires = 2
qml.numpy.random.seed(100)
data = qml.numpy.random.rand(n_wires)
weights = np.ones([n_wires])


@qml.qnode(qml.device("lightning.qubit", wires=n_wires), diff_method="adjoint")
def circuit(weights, data):

    for i in range(n_wires):
        qml.RY(data[i], wires=i)
        qml.RX(weights[i], wires=i)

    return [qml.expval(qml.PauliZ(i)) for i in range(n_wires)]


def loss_fn(weights, data):
    return jnp.array(circuit(weights, data))


print("\n-- Catalyst --")
print(f"Jacobian for arg0:\n{jnp.round(results[0], decimals=5)}")
print(f"Jacobian for arg1:\n{jnp.round(results[1], decimals=5)}")

grad_fn = jax.jacobian(loss_fn, argnums=[0, 1])
results = grad_fn(weights, data)
print("\n-- PL/JAX --")
print(f"Jacobian for arg0:\n{jnp.round(results[0], decimals=5)}")
print(f"Jacobian for arg1:\n{jnp.round(results[1], decimals=5)}")
```

```
-- Catalyst --
Jacobian for arg0:
[[ 0.      -0.14847]
 [ 0.      -0.80908]]
Jacobian for arg1:
[[-0.27937 -0.     ]
 [-0.72026  0.     ]]

-- JAX --
Jacobian for arg0:
[[-0.72026  0.     ]
 [ 0.      -0.80908]]
Jacobian for arg1:
[[-0.27937  0.     ]
 [-0.      -0.14847]]
```
We can see the results being associated to the wrong diff parameters.

Tracking this down through catalyst and the runtime is all consistent. The important handoff point is the [Gradient function](https://github.com/PennyLaneAI/pennylane-lightning/blob/13959918434a06dd4392132c22563765e93f040f/pennylane_lightning/core/simulators/lightning_qubit/catalyst/LightningSimulator.cpp#L623) in the device. I already checked in the debugger that the collected tape looks correct.
The Jacobian first [computed here](https://github.com/PennyLaneAI/pennylane-lightning/blob/13959918434a06dd4392132c22563765e93f040f/pennylane_lightning/core/simulators/lightning_qubit/catalyst/LightningSimulator.cpp#L690) looks as follows:

<img width="360" alt="Image" src="https://github.com/user-attachments/assets/c4216a75-ba3a-4a35-b605-7d1afa3cac43" />

I don't know if this is already wrong or not, but assuming it's correct (core lightning functionality), then the error is likely in the data movement that follows:
https://github.com/PennyLaneAI/pennylane-lightning/blob/13959918434a06dd4392132c22563765e93f040f/pennylane_lightning/core/simulators/lightning_qubit/catalyst/LightningSimulator.cpp#L694-L708
